### PR TITLE
Fixes #979 what will happen during clone and merge for module building blocks

### LINKED ISSUE #979: What will happen during clone and merge for module building blocks
Clone and merge workflows formerly added the result directly to a project. However, in OSMOSES, the resulting building blocks are not top-level members of a project (except Individual and Expression).

### PR DESCRIPTION
**Pull request by rwmcintosh**
There are several related issues surrounding the implicit understanding in MoBi that building blocks are top level members of a project.

To make the clone work for both start value bb's (which belong to modules) and Expression/Individuals (which belong to projects), and Modules, some reorganization of the interaction tasks had to be done. That had some other impacts to - especially in context menus for example

Lots of the code changes are based on that. I tried to highlight spots where new stuff is going on using comments.

SBML importer was formerly loading building blocks to a project, but it now creates a project with one module.

### CODE REVIEW

File: src/MoBi.Presentation/Tasks/Interaction/InteractionTasksForModule.cs [hunk 1]
```diff
@@ -73,5 +81,106 @@ public void AddBuildingBlocksToModule(Module module)
...
+
+      public void RemoveModule(Module module)
+      {
+         _interactionTaskContext.Context.AddToHistory(new RemoveModuleCommand(module)
+            .Run(_interactionTaskContext.Context));
+      }
+
+      public void AddCloneToProject(Module moduleToClone)
+      {
+         var clonedModule = _interactionTaskContext.Context.Clone(moduleToClone);
```
Comment (rwmcintosh) [inline src/MoBi.Presentation/Tasks/Interaction/InteractionTasksForModule.cs:103]: I think the easiest way is to clone the whole module and remove what the user does not want

File: src/MoBi.Presentation/Tasks/Interaction/InteractionTasksForModuleBuildingBlocks.cs [hunk 2]
```diff
@@ -1,134 +0,0 @@
-using System.Collections.Generic;
```
Comment (rwmcintosh) [inline src/MoBi.Presentation/Tasks/Interaction/InteractionTasksForModuleBuildingBlocks.cs:1]: All moved to the InteractionTasksForModule

File: src/MoBi.Presentation/Presenter/CloneBuildingBlocksToModulePresenter.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+   public interface ICloneBuildingBlocksToModulePresenter : IDisposablePresenter
+   {
+      /// <summary>
+      ///    Selectively removes building blocks from <paramref name="clonedModule" /> based on user selection
+      /// </summary>
+      /// <returns>False if the user canceled, otherwise True</returns>
+      bool SelectClonedBuildingBlocks(Module clonedModule);
+   }
+
+   public class CloneBuildingBlocksToModulePresenter : AbstractDisposablePresenter<ICloneBuildingBlocksToModuleView, ICloneBuildingBlocksToModulePresenter>,
```
Comment (rwmcintosh) [inline src/MoBi.Presentation/Presenter/CloneBuildingBlocksToModulePresenter.cs:20]: New presenter for cloning modules

File: src/MoBi.Presentation/Tasks/Interaction/InteractionTasksForPathAndValueEntity.cs [hunk 4]
```diff
@@ -167,17 +178,5 @@ protected IMoBiCommand SetValueWithUnit(TBuilder builder, double? unitValueToBas
       {
          return new PathAndValueEntityValueOrUnitChangedCommand<TBuilder, TBuildingBlock>(builder, unitValueToBaseUnitValue, unit, buildingBlock).Run(Context);
       }
-
-      public override IMoBiCommand Merge(TBuildingBlock buildingBlockToMerge, TBuildingBlock targetBuildingBlock)
-      {
-         // TODO
```
Comment (rwmcintosh) [inline src/MoBi.Presentation/Tasks/Interaction/InteractionTasksForPathAndValueEntity.cs:173]: One TODO down!

File: src/MoBi.Presentation/Tasks/Interaction/AbstractStartValuesTask.cs [hunk 5]
```diff
@@ -271,6 +251,18 @@ public ICommand SetValueOrigin(TBuildingBlock buildingBlock, ValueOrigin valueOr
       }
 
       public abstract bool CanResolve(TBuildingBlock buildingBlock, TStartValue startValue);
+      
+      public ICommand CloneAndAddToParent(TBuildingBlock buildingBlockToClone, Module parentModule)
```
Comment (rwmcintosh) [inline src/MoBi.Presentation/Tasks/Interaction/AbstractStartValuesTask.cs:255]: New clone method for start values

File: src/MoBi.UI/Views/BaseModuleContentView.cs [hunk 6]
```diff
@@ -10,11 +8,12 @@
 
 namespace MoBi.UI.Views
 {
-   public partial class CreateModuleView : BaseModalView, ICreateModuleView
+   public abstract partial class BaseModuleContentView<TDTO> : BaseModalView
```
Comment (rwmcintosh) [inline src/MoBi.UI/Views/BaseModuleContentView.cs:11]: This view is pretty much identical for various presenters that enable users to select certain building blocks of a module

Create, Add, Clone as of now. I refactored it with a common base and minimal concrete implementation details

File: src/MoBi.Core/Commands/RemoveBuildingBlockCommand.cs [hunk 7]
```diff
@@ -36,7 +36,7 @@ public override void RestoreExecutionData(IMoBiContext context)
 
       protected override ICommand<IMoBiContext> GetInverseCommand(IMoBiContext context)
       {
-         return new AddBuildingBlockCommand<T>(_buildingBlock).AsInverseFor(this);
+         return new AddProjectBuildingBlockCommand<T>(_buildingBlock).AsInverseFor(this);
```
Comment (msevestre) [inline src/MoBi.Core/Commands/RemoveBuildingBlockCommand.cs:39]: AddBuildingBlockToProject?
and RemoveBuildingBlockFromProject?

File: src/MoBi.Presentation/MenusAndBars/ContextMenus/ContextMenuForProjectBuildingBlock.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+   public class ContextMenuForProjectBuildingBlock<TBuildingBlock> : ContextMenuForBuildingBlock<TBuildingBlock> where TBuildingBlock : class, IBuildingBlock
+   {
+      public ContextMenuForProjectBuildingBlock(IMoBiContext context, IObjectTypeResolver objectTypeResolver, IContainer container) : base(context, objectTypeResolver, container)
+      {
+      }
+
+      public override IContextMenu InitializeWith(ObjectBaseDTO dto, IPresenter presenter)
+      {
+         var buildingBlock = _context.Get<TBuildingBlock>(dto.Id);
+         base.InitializeWith(dto, presenter);
```
Comment (msevestre) [inline src/MoBi.Presentation/MenusAndBars/ContextMenus/ContextMenuForProjectBuildingBlock.cs:26]: weird to have base not first. Why?

File: src/MoBi.Presentation/Presenter/Main/ModuleExplorerPresenter.cs [hunk 9]
```diff
@@ -108,8 +108,16 @@ public override bool CanDrag(ITreeNode node)
 
       public int OrderingComparisonFor(ITreeNode<IWithName> node1, ITreeNode<IWithName> node2)
       {
+         if (nodeIsStartValueFolderNode(node1) && nodeIsStartValueFolderNode(node2))
+            return nodeIsMoleculeStartValuesNode(node1) ? -1 : 1;
+
+         if (nodeIsStartValueFolderNode(node1))
+            return 1;
+         if (nodeIsStartValueFolderNode(node2))
```
Comment (msevestre) [inline src/MoBi.Presentation/Presenter/Main/ModuleExplorerPresenter.cs:116]: run resharper

File: src/MoBi.Presentation/Tasks/Interaction/AbstractStartValuesTask.cs (hunk 5 above)
Comment (msevestre) [reply src/MoBi.Presentation/Tasks/Interaction/AbstractStartValuesTask.cs:255]: why not xxAddToModule?

File: src/MoBi.Presentation/Tasks/Interaction/BuildingBlockTaskRetriever.cs [hunk 10]
```diff
@@ -45,62 +45,64 @@ public IInteractionTasksForBuildingBlock TaskFor(IBuildingBlock buildingBlock)
...
       {
-         _task = _container.Resolve<IInteractionTasksForBuildingBlock<TBuildingBlock>>();
+         _task = _container.Resolve<IInteractionTasksForBuildingBlock<TParent, TBuildingBlock>>();
       }
- 
+
       public void Visit(MoleculeBuildingBlock objToVisit)
       {
-         retrieveTask(objToVisit);
+         retrieveTask<Module, MoleculeBuildingBlock>(objToVisit);
```
Comment (msevestre) [inline src/MoBi.Presentation/Tasks/Interaction/BuildingBlockTaskRetriever.cs:60]: why do you specifrcy Module all the time here? Cannot the line 55 use Module instead?

File: src/MoBi.Presentation/Tasks/Interaction/InteractionTasksForModule.cs [hunk 11]
```diff
@@ -73,5 +81,106 @@ public void AddBuildingBlocksToModule(Module module)
...
+         {
+            if (presenter.SelectClonedBuildingBlocks(clonedModule) == false)
+               return;
+         }
+         
+         _interactionTaskContext.Context.AddToHistory(GetAddCommand(clonedModule, _interactionTaskContext.Context.CurrentProject, null)
+            .Run(_interactionTaskContext.Context));
+      }
+
+      private void loadBuildingBlocksToModuleBase(Module module, Func<string> getFilename)
```
Comment (msevestre) [inline src/MoBi.Presentation/Tasks/Interaction/InteractionTasksForModule.cs:115]: why not loadBuildingBlocksToModule? Base feels weird in the naming.

**Pull request merged**